discovery/src/main/java/io/airlift/discovery/client/DiscoveryBinder.java

```diff
@@ -104,15 +117,90 @@ public void bindHttpSelector(ServiceType serviceType)
     }
 
+    public static class HttpServiceAnnouncementBuilder
+    {
+        private final MapBinder<String, String> propertiesBinder;
+
+        public HttpServiceAnnouncementBuilder(MapBinder<String, String> propertiesBinder)
+        {
+            this.propertiesBinder = propertiesBinder;
+        }
+
+        public PropertyBinding bindProperty(String key)
+        {
+            requireNonNull(key, "key is null");
+            return new PropertyBinding()
+            {
+                @Override
+                public HttpServiceAnnouncementBuilder toProvider(Provider<String> provider)
+                {
+                    propertiesBinder.addBinding(key).toProvider(provider);
+                    return HttpServiceAnnouncementBuilder.this;
+                }
+
+                @Override
+                public HttpServiceAnnouncementBuilder toProvider(Key<? extends Provider<String>> providerKey)
+                {
+                    propertiesBinder.addBinding(key).toProvider(providerKey);
+                    return HttpServiceAnnouncementBuilder.this;
+                }
+            };
+        }
+
+        @CanIgnoreReturnValue
+        public HttpServiceAnnouncementBuilder addProperty(String key, String value)
+        {
+            bindProperty(key).toInstance(value);
+            return this;
+        }
+
+        @CanIgnoreReturnValue
+        public HttpServiceAnnouncementBuilder addProperties(Map<String, String> properties)
+        {
+            properties.forEach(this::addProperty);
+            return this;
+        }
+
+        public interface PropertyBinding
+        {
+            @CanIgnoreReturnValue
+            default HttpServiceAnnouncementBuilder toInstance(String value)
+            {
+                return toProvider(Providers.of(requireNonNull(value, "value is null")));
+            }
+
+            @CanIgnoreReturnValue
+            HttpServiceAnnouncementBuilder toProvider(Provider<String> provider);
+
+            @CanIgnoreReturnValue
+            default HttpServiceAnnouncementBuilder toProvider(Class<? extends Provider<String>> providerType)
+            {
+                return toProvider(Key.get(providerType));
+            }
+
+            @CanIgnoreReturnValue
+            HttpServiceAnnouncementBuilder toProvider(Key<? extends Provider<String>> providerKey);
+        }
+    }
+
     static class HttpAnnouncementProvider
             implements Provider<ServiceAnnouncement>
     {
-        private final ServiceAnnouncementBuilder builder;
+        private final String type;
+        private final Annotation annotation;
+        private Injector injector;
         private AnnouncementHttpServerInfo httpServerInfo;
 
-        public HttpAnnouncementProvider(ServiceAnnouncementBuilder serviceAnnouncementBuilder)
+        public HttpAnnouncementProvider(String type, Annotation annotation)
```

Comment:
Add `equals` and `hashCode` based on `type` and `annotation` so that Guice can de-dupe if the provider is bound multiple times.

Reply:
HttpAnnouncementProvider is pre-existing and didn't have equals/hashcode
it feels it would be separate commit. since i don't fully understand the rationale, can this be a follow-up?

Reply:
Not having this will prevent properties from being provided from multiple locations. It's needed to allow
```java
discoveryBinder(binder).bindHttpAnnouncement("apple").addProperty("aaa", "aaa");
discoveryBinder(binder).bindHttpAnnouncement("apple").addProperty("xxx", "xxx");
```
But we could add this later if needed.